zkevm-circuits/src/evm_circuit/execution/error_code_store.rs

```diff
@@ -0,0 +1,311 @@
+    };
+
+    use lazy_static::lazy_static;
+    use mock::{eth, TestContext, MOCK_ACCOUNTS};
+
+    use crate::test_util::CircuitTestBuilder;
+
+    lazy_static! {
+        static ref CALLER_ADDRESS: Address = address!("0x00bbccddee000000000000000000000000002400");
+    }
+
+    const MAXCODESIZE: u64 = 0x6000u64;
+
+    fn run_test_circuits(ctx: TestContext<2, 1>) {
+        CircuitTestBuilder::new_from_test_ctx(ctx)
+            .params(FixedCParams {
+                max_rws: 4500,
+                ..FixedCParams::default()
+            })
+            .run();
+    }
+
+    fn initialization_bytecode(is_oog: bool) -> Bytecode {
+        let memory_bytes = [0x60; 10];
+        let memory_value = Word::from_big_endian(&memory_bytes);
+        let code_len = if is_oog { 0 } else { MAXCODESIZE + 1 };
+
+        let mut code = bytecode! {
+            PUSH10(memory_value)
+            PUSH32(code_len)
+            MSTORE
+            PUSH2(if is_oog { 5 } else { code_len}) // length to copy
+            PUSH2(32u64 - u64::try_from(memory_bytes.len()).unwrap()) // offset
+            //PUSH2(0x00) // offset
+
+        };
+        code.write_op(OpcodeId::RETURN);
+
+        code
+    }
+
+    fn creator_bytecode(initialization_bytecode: Bytecode, is_create2: bool) -> Bytecode {
+        let initialization_bytes = initialization_bytecode.code();
+        let mut code = Bytecode::default();
+
+        // construct maxcodesize + 1 memory bytes
+        let code_creator: Vec<u8> = initialization_bytes
+            .to_vec()
+            .iter()
+            .cloned()
+            .chain(0u8..((32 - initialization_bytes.len() % 32) as u8))
+            .collect();
+        for (index, word) in code_creator.chunks(32).enumerate() {
+            code.push(32, Word::from_big_endian(word));
+            code.push(32, Word::from(index * 32));
+            code.write_op(OpcodeId::MSTORE);
+        }
+
+        if is_create2 {
+            code.append(&bytecode! {PUSH1(45)}); // salt;
+        }
+        code.append(&bytecode! {
+            PUSH32(initialization_bytes.len()) // size
+            PUSH2(0x00) // offset
+            PUSH2(23414) // value
+        });
+        code.write_op(if is_create2 {
+            OpcodeId::CREATE2
+        } else {
+            OpcodeId::CREATE
+        });
+        code.append(&bytecode! {
+            PUSH1(0)
+            PUSH1(0)
+            RETURN
+        });
+
+        code
+    }
+
+    fn test_context(caller: Account, is_oog: bool) -> TestContext<2, 1> {
+        TestContext::new(
+            None,
+            |accs| {
+                accs[0]
+                    .address(address!("0x000000000000000000000000000000000000cafe"))
+                    .balance(eth(10));
+                accs[1].account(&caller);
+            },
+            |mut txs, accs| {
+                txs[0]
+                    .from(accs[0].address)
+                    .to(accs[1].address)
+                    .gas(if is_oog {
```

Comment:
It'll be great if you could add some comment to explain how the numbers comes from.